#12 Service class

Charlottehenriksson opened the pull request:
### Related Issue(s)
#12 


### Description
Changed a typo in repository folder class name,
Service class for user and messages created, dont know if its needed to add a set privacy on messages in service layer

Ebyrdeu commented on src/main/java/dev/ebrydeu/spring_boot_library/services/MessageService.java:
```diff
@@ -0,0 +1,66 @@
+import java.util.stream.Collectors;
+
+@Service
+    @Transactional
```
Didn't watch Martin latest lecturers. What propose of `@Transactional` in service Layer ?

Charlottehenriksson replied:
I believe I read it here: https://medium.com/@bubu.tripathy/implementing-transactions-in-a-spring-boot-application-bc6b33e88557 
I dont remember Martin using it in his cat project :) But if its not needed for this project I will remove it :D

ValentinaSukonina replied:
By annotating methods with @Transactional we specify that the operations within those methods should be executed within a single transaction boundary. This means that either all the operations succeed and are committed to the database, or in case of any failure, all changes are rolled back (undone), ensuring data consistency.

Ebyrdeu replied:
You can leave it. I was just curious.

Ebyrdeu commented on src/main/java/dev/ebrydeu/spring_boot_library/services/MessageService.java:
```diff
@@ -0,0 +1,66 @@
+
+@Service
+    @Transactional
+    public class MessageService {
+        MessageRepository messageRepository;
+
+        @Autowired
+        public MessageService(MessageRepository messageRepository) {
+            this.messageRepository = messageRepository;
+        }
```
One thing about `@Autowired` is that it simply annotates that the Dependency Injection (DI) being used is a Bean. In modern Spring Boot, you don't need to declare `@Autowired` if your DI is considered a Bean. But how do you determine if it's a bean or not? If you have IntelliJ Ultimate, it will show a gutter icon (a small bean icon on the left) for beans. But how can you detect this without the help of IntelliJ? It's quite simple but perhaps tedious. You just need to memorize the most common annotations, such as `@Service`, `@Repository`,`@RestController`, `@Controller` .

So in our case, if you open `@Service`, you would see the `@Component` annotation, which is kind of a sibling. What it does is it automatically finds and does `automatic wiring` (essentially the same as what `@Autowired` does, but you don't need to annotate).

But in terms of when we actually need to use `@Autowired`, it's quite simple,  `in tests`.

Nevertheless, even with this, we use Lombok and we don't need a constructor here. We can use just one annotation, which is `@RequiredArgsConstructor`, which does the same job, but only when we have `final` on the field. Also change for `UserService`

```suggestion
@RequiredArgsConstructor
    public class MessageService {
       final MessageRepository messageRepository;
```

Charlottehenriksson replied:
Its removed, thanks for easy to understand info :)

Ebyrdeu commented on src/main/java/dev/ebrydeu/spring_boot_library/services/UserService.java:
```diff
@@ -0,0 +1,76 @@
+    }
+    public UserDto getUserById(Long userId) {
+        Optional<User> userOptional = userRepository.findById(userId);
+        if (userOptional.isEmpty()) {
+            throw new CustomExceptions.NotFoundException("User not found with id: " + userId);
+        }
```
To maintain consistency, do the same as in `Message`, or alternatively, adjust `Message` to match the "User not found" convention.

```Java 
Message message = messageRepository.findById(id).orElseThrow(() -> new CustomExceptions.NotFoundException("Message not found with id: " + id));

Charlottehenriksson commented:
public UserDto createUser(UserDto userDto) {
        if (!userRepository.findByEmail(userDto.email()).isEmpty()) {
            throw new CustomExceptions.EmailAlreadyExistsException("Email already exists: " + userDto.email());
        }
        User user = UserDto.map(userDto);
        User savedUser = userRepository.save(user);
        return UserDto.map(savedUser);
    }
    
    Im having a bit of trouble here, the other exception throws are similar to the messageService class but this one im not sure how to change so it will be similiar to the messageService class :)

Ebyrdeu commented:
The issue stems from the implementation of the `Repository` interface. For instance, in your case, the `findByEmail` method returns all users with that email (which doesn't quite make sense if you think about it we have unique email.  I overlooked this previously, so my apologies for that). 

However, what you need to do is go to the `UserRepository` and make the necessary changes.

```Java
@Repository
public interface UserRepository extends ListCrudRepository<User, Long> {
    List<User> findByProfileName(String profileName);
    List<User> findByFirstName(String firstName);
    List<User> findByLastName(String lastName);
    Optional<User> findByEmail(String email);
}

ValentinaSukonina commented on src/main/java/dev/ebrydeu/spring_boot_library/services/MessageService.java:
```diff
@@ -0,0 +1,66 @@
+                    .collect(Collectors.toList());
+        }
+
+        public MessageDto saveMessage(MessageDto messageDto) {
```
I think you need to add @CacheEvict before methods that will modify our database (such as save, edit and so on, otherwise the old information will remained in the Cache

Charlottehenriksson replied:
Great thanks, its added now!

ValentinaSukonina reviewed (changes requested):
Goot work, but some small addition I think will be necessary ("I think you need to add @CacheEvict before methods that will modify our database (such as save, edit and so on, otherwise the old information will remained in the Cachesome small changes")

Charlottehenriksson commented:
> The issue stems from the implementation of the `Repository` interface. For instance, in your case, the `findByEmail` method returns all users with that email (which doesn't quite make sense if you think about it we have unique email. I overlooked this previously, so my apologies for that).
> 
> However, what you need to do is go to the `UserRepository` and make the necessary changes.
> 
> ```java
> @Repository
> public interface UserRepository extends ListCrudRepository<User, Long> {
>     List<User> findByProfileName(String profileName);
>     List<User> findByFirstName(String firstName);
>     List<User> findByLastName(String lastName);
>     Optional<User> findByEmail(String email);
> }
> ```

I have updated the repository! And I have changed createUser in the code, hopefully its correct :)

ValentinaSukonina reviewed (dismissed):
Good work!

ValentinaSukonina commented:
The use of Optional<User> conveys the possibility of a nullable result explicitly, and it's beneficial for methods that may or may not find a matching object. On the other hand, returning a List<User> implies that there could be multiple matches (although it's unusual for email addresses to have multiple users, we however did not specify this parameter as unique), and if no matches are found, it returns an empty list. Eventually it is up to us how we want to handle it. I have no strong preferences and gladly accept your changes. Test for UserRepository is now modified accordingly, plus test for non-existing user is added.

Ebyrdeu commented:
You forgot - `@RequiredArgsConstructor` in `UserService`

Charlottehenriksson commented:
> You forgot - `@RequiredArgsConstructor` in `UserService`

Fixed it! Good catch :)

ValentinaSukonina reviewed (dismissed):
Looks good

ajohnsson95 reviewed (commented):
looks great!

Ebyrdeu commented:
Sonnar compain that you don't use return value  of ` userRepository.findById(userId)`
```Java 
public void deleteUser(Long userId) {
        userRepository.findById(userId)
                .orElseThrow(() -> new CustomExceptions.NotFoundException("User not found with id: " + userId));`
```

```Java 
    @CacheEvict("users")
    public void deleteUser(Long userId) {
        User  user = userRepository.findById(userId)
                .orElseThrow(() -> new CustomExceptions.NotFoundException("User not found with id: " + userId));

        userRepository.deleteById(user.getId());
    }

Charlottehenriksson commented:
> Sonnar compain that you don't use return value of ` userRepository.findById(userId)`
> 
> ```java
> public void deleteUser(Long userId) {
>         userRepository.findById(userId)
>                 .orElseThrow(() -> new CustomExceptions.NotFoundException("User not found with id: " + userId));`
> ```
> 
> ```java
>     @CacheEvict("users")
>     public void deleteUser(Long userId) {
>         User  user = userRepository.findById(userId)
>                 .orElseThrow(() -> new CustomExceptions.NotFoundException("User not found with id: " + userId));
> 
>         userRepository.deleteById(user.getId());
>     }
> ```

Fixed the return value and all checks passed :)

ValentinaSukonina reviewed (commented):
Goot work with updating delete methods. Sonar still complains about using Stream.collect(Collectors.toList() instead of Stream.toList().
The toList() method was introduced in Java 16 as a shorthand for collect(Collectors.toList()), providing a more concise way to collect elements into a list. It offers the same behavior and performance characteristics as collect(Collectors.toList()).
In summary, if we are using Java 16 or later, toList() is preferred for its brevity and readability. Otherwise, using collect(Collectors.toList()) is perfectly fine. Any suggestions from others? Otherwise it looks good!

ValentinaSukonina commented on src/main/java/dev/ebrydeu/spring_boot_library/services/UserService.java:
```diff
@@ -0,0 +1,79 @@
+                .map(UserDto::map)
+                .collect(Collectors.toList());
+    }
+    @Cacheable("users")
```
Version that Sonar accept without complains is:
    @Cacheable("users")
    public List<UserDto> findByProfileName1(String profileName) {
        List<User> users = userRepository.findByProfileName(profileName);
        return users.stream()
                .map(UserDto::map)
                .toList();
    }

Charlottehenriksson commented:
> Goot work with updating delete methods. Sonar still complains about using Stream.collect(Collectors.toList() instead of Stream.toList(). The toList() method was introduced in Java 16 as a shorthand for collect(Collectors.toList()), providing a more concise way to collect elements into a list. It offers the same behavior and performance characteristics as collect(Collectors.toList()). In summary, if we are using Java 16 or later, toList() is preferred for its brevity and readability. Otherwise, using collect(Collectors.toList()) is perfectly fine. Any suggestions from others? Otherwise it looks good!

Great explanation! The methods using collectors.toList() are now updated :D

ValentinaSukonina reviewed (approved):
Good work with all updates and changes
